File: apps/worker/worker_python/pipeline/embedding_contract.py
"""The fixed embedding contract used by indexing and evaluation."""

from __future__ import annotations

import json
import logging
import math
import os
import struct
from collections.abc import Mapping
from dataclasses import dataclass

EMBEDDING_DIMENSIONS = 1536
logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class EmbeddingConfig:
    provider: str
    model: str


def embedding_diagnostic(config: EmbeddingConfig | None = None, **details) -> dict:
    return {
        "event": "embedding.validation",
        **({"provider": config.provider, "model": config.model} if config else {}),
        "expected_dimensions": EMBEDDING_DIMENSIONS,
        **details,
    }


class EmbeddingContractError(RuntimeError):
    def __init__(self, reason: str, message: str, config: EmbeddingConfig | None = None, **details):
        super().__init__(message)
        self.reason = reason
        logger.error(json.dumps(embedding_diagnostic(config, reason=reason, **details)))


def resolve_embedding_config(environ: Mapping[str, str] | None = None) -> EmbeddingConfig:
    env = os.environ if environ is None else environ
    provider = env.get("EMBEDDING_PROVIDER", "").strip().lower() or "openai"
    if provider not in {"openai", "ollama"}:
        raise EmbeddingContractError("EMBEDDING_CONFIG_INVALID", "Unsupported embedding provider. Configure openai or ollama.")
    model = env.get("EMBEDDING_MODEL", "").strip() or (
        "text-embedding-3-small" if provider == "openai" else ""
    )
    if not model:
        raise EmbeddingContractError("EMBEDDING_CONFIG_INVALID", "EMBEDDING_MODEL is required when EMBEDDING_PROVIDER=ollama.")
    return EmbeddingConfig(provider, model)


def invalid_output(config: EmbeddingConfig) -> EmbeddingContractError:
    return EmbeddingContractError("EMBEDDING_OUTPUT_INVALID", "Embedding response is incompatible with the 1536-dimensional storage contract.", config)


def validate_embedding(value: object, config: EmbeddingConfig) -> list[float]:
    # pgvector stores float32: reject overflow and vectors that round to all zeros.
    try:
        valid = isinstance(value, list) and len(value) == EMBEDDING_DIMENSIONS and all(
            type(v) in (int, float) and math.isfinite(v) for v in value
        )
        rounded = [struct.unpack("f", struct.pack("f", v))[0] for v in value] if valid else []
        valid = valid and all(math.isfinite(v) for v in rounded) and any(v != 0 for v in rounded)
    except (OverflowError, struct.error):
        valid = False
    if not valid:
        raise EmbeddingContractError(
            "EMBEDDING_OUTPUT_INVALID", "Embedding response is incompatible with the 1536-dimensional storage contract.",
            config, actual_dimensions=len(value) if isinstance(value, list) else None,
        )
    return value

File: apps/api/src/lib/embedding-contract.ts
import { DEFAULT_EMBEDDING_MODEL, LlmConfigurationError, LlmProviderError } from "./openai";
import { EMBEDDING_DIMENSIONS } from "./embedding-dimensions";

export { EMBEDDING_DIMENSIONS };

export type EmbeddingConfig = { provider: "openai" | "ollama"; model: string };
export type EmbeddingSettings = { EMBEDDING_PROVIDER?: string; EMBEDDING_MODEL?: string };
export type EmbeddingReason =
  | "EMBEDDING_CONFIG_INVALID"
  | "EMBEDDING_SCHEMA_MISMATCH"
  | "EMBEDDING_OUTPUT_INVALID"
  | "EMBEDDING_DATA_INVALID";

export function embeddingDiagnostic(
  config: EmbeddingConfig | undefined,
  details: { reason?: EmbeddingReason; db_dimensions?: number | null; actual_dimensions?: number | null } = {},
) {
  return { event: "embedding.validation", ...config, expected_dimensions: EMBEDDING_DIMENSIONS, ...details };
}

export class EmbeddingConfigurationError extends LlmConfigurationError {
  constructor(
    readonly reason: "EMBEDDING_CONFIG_INVALID" | "EMBEDDING_SCHEMA_MISMATCH",
    message: string,
    config?: EmbeddingConfig,
    dbDimensions?: number | null,
  ) {
    super(message);
    console.error(JSON.stringify(embeddingDiagnostic(config, { reason, db_dimensions: dbDimensions })));
  }
}

export class EmbeddingValidationError extends LlmProviderError {
  constructor(
    readonly reason: "EMBEDDING_OUTPUT_INVALID" | "EMBEDDING_DATA_INVALID",
    config?: EmbeddingConfig,
    actualDimensions?: number | null,
  ) {
    super("Embedding data is incompatible. Ask an administrator to check the model and rebuild the affected embeddings.");
    console.error(JSON.stringify(embeddingDiagnostic(config, { reason, actual_dimensions: actualDimensions })));
  }
}

export function resolveEmbeddingConfig(env: EmbeddingSettings): EmbeddingConfig {
  const provider = env.EMBEDDING_PROVIDER?.trim().toLowerCase() || "openai";
  if (provider !== "openai" && provider !== "ollama") {
    throw new EmbeddingConfigurationError("EMBEDDING_CONFIG_INVALID", "Unsupported embedding provider. Configure openai or ollama.");
  }
  const model = env.EMBEDDING_MODEL?.trim() || (provider === "openai" ? DEFAULT_EMBEDDING_MODEL : "");
  if (!model) {
    throw new EmbeddingConfigurationError("EMBEDDING_CONFIG_INVALID", "EMBEDDING_MODEL is required when EMBEDDING_PROVIDER=ollama.");
  }
  return { provider, model };
}

export function validateEmbedding(
  value: unknown,
  reason: "EMBEDDING_OUTPUT_INVALID" | "EMBEDDING_DATA_INVALID" = "EMBEDDING_OUTPUT_INVALID",
  config?: EmbeddingConfig,
): number[] {
  if (
    !Array.isArray(value) || value.length !== EMBEDDING_DIMENSIONS ||
    !value.every((v: unknown) => typeof v === "number" && Number.isFinite(v) && Number.isFinite(Math.fround(v))) ||
    !value.some((v: number) => Math.fround(v) !== 0)
  ) {
    throw new EmbeddingValidationError(reason, config, Array.isArray(value) ? value.length : null);
  }
  return value;
}

/** Empty arrays represent an ungenerated concept; malformed stored data does not. */
export function parseStoredEmbedding(value: unknown): number[] {
  let parsed: unknown = value;
  if (typeof value === "string") {
    try { parsed = JSON.parse(value); }
    catch { throw new EmbeddingValidationError("EMBEDDING_DATA_INVALID"); }
  }
  if (parsed == null || (Array.isArray(parsed) && parsed.length === 0)) return [];
  return validateEmbedding(parsed, "EMBEDDING_DATA_INVALID");
}
